org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java

```diff
@@ -0,0 +1,187 @@
+ */
+package org.eclipse.xtext.ui.testing.util;
+
+import java.io.IOException;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.xtext.ui.MarkerTypes;
+
+import com.google.common.base.Strings;
+
+/**
+ * @author Moritz Eysholdt - Initial contribution and API
+ */
+public class AnnotatedTextToString {
+
+	protected static class CommentedRegion {
+
+		private final int start;
+
+		private final int end;
+
+		private final String text;
+
+		public CommentedRegion(int start, int end, String text) {
+			this.start = start;
+			this.end = end;
+			this.text = text;
+		}
+	}
+
+	private String contents;
+
+	private IFile file;
+
+	private List<CommentedRegion> regions = new ArrayList<>();
+
+	private List<String> emptyMessages = new ArrayList<>();
+
+	public String getContents() {
+		try {
+			if (contents != null) {
+				return contents;
+			}
+
+			if (file != null) {
+				return IResourcesSetupUtil.fileToString(file);
+			}
+
+			return null;
+		} catch (CoreException | IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	public void setContents(String contents) {
+		this.contents = contents;
+	}
+
+	public IFile getFile() {
+		return file;
+	}
+
+	public void setFile(IFile file) {
+		this.file = file;
+	}
+
+	public List<CommentedRegion> getRegions() {
+		return regions;
+	}
+
+	public List<String> getEmptyMessage() {
+		return emptyMessages;
+	}
+
+	public AnnotatedTextToString withFile(IFile file) {
+		this.file = file;
+		return this;
+	}
+
+	public AnnotatedTextToString withMarkers(IMarker[] markers) {
+		return withMarkers(markers, IMarker.MESSAGE);
+	}
+
+	public AnnotatedTextToString withMarkers(IMarker[] markers, String... attributes) {
+		try {
+			for (IMarker marker : markers) {
+				List<String> transformed = new ArrayList<String>(attributes.length);
+				for (String attribute : attributes) {
+					transformed.add(attribute + "=" + marker.getAttribute(attribute));
+				}
+
+				String comment = transformed.stream().collect(Collectors.joining(", "));
+				Integer offset = (Integer) marker.getAttribute(IMarker.CHAR_START);
+				Integer length = (Integer) marker.getAttribute(IMarker.CHAR_END);
+				regions.add(new CommentedRegion(offset, length, comment));
+			}
+		} catch (CoreException e) {
+			// Log the exception?
+		}
+		emptyMessages.add("(no markers found)");
+		return this;
+	}
+
+	public AnnotatedTextToString withMarkersFromFile(String type, String... attributes) {
+		try {
+			return withMarkers(file.findMarkers(type, true, IResource.DEPTH_INFINITE), attributes);
+		} catch (CoreException e) {
+			// Log the exception and return null or throw RuntimeException?
```

Comment:
I believe we should log a message here, and return a `null` value. @szarnekow what do you think?